crates/widgets/src/progress_bar.rs

```diff
@@ -0,0 +1,98 @@
+use crate::prelude::*;
+
+static RANGE_MIN: f64 = 0.0;
+static RANGE_MAX: f64 = 1.0;
+
+#[derive(Default, AsAny)]
+struct BarState {}
+
+impl State for BarState {
+    fn init(&mut self, _: &mut Registry, ctx: &mut Context) {
+        ctx.child_from_index(0)
+            .get_mut::<Constraint>("constraint")
+            .set_width(0.1);
+    }
+
+    fn update(&mut self, _: &mut Registry, ctx: &mut Context<'_>) {
+        let val = ctx.widget().clone_or_default::<f64>("val");
+        let new_width: f64;
+
+        if val == RANGE_MIN {
+            new_width = 0.1;
+        } else {
+            let max_width = ctx.widget().get::<Rectangle>("bounds").width();
```

Comment:
Suggestion: Extract the calculation and write a unit test.

e.g.

https://github.com/redox-os/orbtk/blob/928a4c1f39de7ed158d7a0feb1ef5b16bdc40ce5/crates/widgets/src/slider.rs#L257

https://github.com/redox-os/orbtk/blob/928a4c1f39de7ed158d7a0feb1ef5b16bdc40ce5/crates/widgets/src/slider.rs#L290